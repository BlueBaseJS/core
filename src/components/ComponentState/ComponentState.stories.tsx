import { BlueRainType } from '../../index';
import { bluerainStory } from '../../../storybook/bluerain';
import React from 'react';
import storiesOf from '../../../storybook/storiesOf';

storiesOf('ComponentState', module)
	.add('Only title', bluerainStory((props: { bluerain: BlueRainType }) => {

		const BR = props.bluerain;
		const ComponentState = BR.Components.get('ComponentState');

		return <ComponentState title="A simple state" />;
	}))
	.add('title with titleStyle prop', bluerainStory((props: { bluerain: BlueRainType }) => {

		const BR = props.bluerain;
		const ComponentState = BR.Components.get('ComponentState');
		const Style = { color: 'green', fontSize: 20 };

		return <ComponentState title="A simple state" titleStyle={BR.Utils.createStyleSheet(Style)} />;
	}))
	.add('Only description', bluerainStory((props: { bluerain: BlueRainType }) => {

		const BR = props.bluerain;
		const ComponentState = BR.Components.get('ComponentState');

		return <ComponentState description="A simple description" />;
	}))
	.add('description with Styles ', bluerainStory((props: { bluerain: BlueRainType }) => {

		const BR = props.bluerain;
		const ComponentState = BR.Components.get('ComponentState');
		const Style={ color:'green',fontSize:20 };

		return <ComponentState description="A simple description"  descriptionStyle={Style} />;
	}))
	.add('Only buttonTitle', bluerainStory((props: { bluerain: BlueRainType }) => {

		const BR = props.bluerain;
		const ComponentState = BR.Components.get('ComponentState');

		return <ComponentState buttonTitle="Bluerain"  />;
	}))
	.add('Button with buttonOnPress Prop', bluerainStory((props: { bluerain: BlueRainType }) => {
		const Handler=() => console.log('clicked');

		const BR = props.bluerain;
		const ComponentState = BR.Components.get('ComponentState');

		return <ComponentState   buttonTitle="Bluerain" buttonOnPress={Handler}  />;
	}))
	.add('Button with buttonStyle prop', bluerainStory((props: { bluerain: BlueRainType }) => {
		const BR = props.bluerain;
		const ComponentState = BR.Components.get('ComponentState');
		const Style={ color:'white',backgroundColor:'red',width:100,height:50,borderRadius:8, textAlign:'center' };

		return <ComponentState  buttonTitle="Bluerain"  buttonStyle={BR.Utils.createStyleSheet(Style)}  />;

	}))
	.add('Button Component', bluerainStory((props: { bluerain: BlueRainType }) => {

		const Button=() =>  <button>ButtonComponent</button>;
		const BR = props.bluerain;
		const ComponentState = BR.Components.get('ComponentState');
		// const Style={ color:'white',backgroundColor:'red',width:100,height:50,borderRadius:8, textAlign:'center' };

		return <ComponentState  button={Button}  />;
	}))
	.add('imageComponent prop', bluerainStory((props: { bluerain: BlueRainType }) => {

		const BR = props.bluerain;
		const Image = BR.Components.get('Image');
		const Style = { width: 200, height: 200 };
		// tslint:disable-next-line:max-line-length
		const ImageComponent = () => <Image source="https://lh3.googleusercontent.com/0-BzaWtxoAnsBjQ_wzUcKxyF07XE7v2Kkg1ogPVUdzmQpvaz118uHQEGU6BdtzJuzfo=h1264" imageStyle={BR.Utils.createStyleSheet(Style)} />;
		const ComponentState = BR.Components.get('ComponentState');
		return <ComponentState image={ImageComponent} />;
	}))

	.add('imageSource prop', bluerainStory((props: { bluerain: BlueRainType }) => {

		const BR = props.bluerain;
		// const Image = BR.Components.get('Image');
		const Style={ width:200,height:200 };
		const ComponentState = BR.Components.get('ComponentState');
		// tslint:disable-next-line:max-line-length
		return <ComponentState  imageSource="https://lh3.googleusercontent.com/0-BzaWtxoAnsBjQ_wzUcKxyF07XE7v2Kkg1ogPVUdzmQpvaz118uHQEGU6BdtzJuzfo=h1264"  imageStyle={BR.Utils.createStyleSheet(Style)} />;
	}));

// storiesOf('ComponentState', module).add('Simple', () => <div>hello </div>);
