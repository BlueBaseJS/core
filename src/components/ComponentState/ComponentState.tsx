import React from 'react';
import { ViewProperties, ImageStyle, ViewStyle, TextStyle } from '@blueeast/bluerain-ui-interfaces';
import { withBlueRain, BlueRainType } from '../../index';

export interface ComponentStateProps extends ViewProperties {
	/**
	 * Image Component, if provided, imageSource will be ignored
	 */
	image?: React.ComponentType;

	/**
	 * Image styles
	 */
	imageStyle?: ImageStyle;

	/**
	 * Image source
	 */
	imageSource?: string;

	/**
	 * Title text
	 */
	title: string;

	/**
	 * Title style
	 */
	titleStyle?: TextStyle;

	/**
	 * Description Text
	 */
	description: string;

	/**
	 * Description style
	 */
	descriptionStyle?: TextStyle;

	/**
	 * Button Component, if provided, other button props will be ignored
	 */
	button?: React.ComponentType;

	/**
	 * Button title
	 */
	buttonTitle?: string;

	/**
	 * Button styles
	 */
	buttonStyle?: ViewStyle;

	/**
	 * Button onPress handler
	 */
	buttonOnPress?: Function;

	style:ViewStyle
}

const ComponentState = (props: ComponentStateProps & { bluerain: BlueRainType }) => {
	const {
		image: ImageComponent,
		imageSource,
		title,
		description,
		button: ButtonComponent,
		buttonTitle,
		buttonOnPress,
		bluerain: BR
	} = props;
	const View = BR.Components.get('View');
	const Text = BR.Components.get('Text');
	const Image = BR.Components.get('Image');
	const Button = BR.Components.get('Button');

	const style = {
		alignItems: 'center',
		padding: 16,
	};

	const stylesheet = [
		BR.Utils.createStyleSheet(style),
		BR.Utils.createStyleSheet(props.style || {})
	];

	const titleStyle = BR.Utils.createStyleSheet(props.titleStyle || {});
	const descriptionStyle = BR.Utils.createStyleSheet(props.descriptionStyle || {});

	// Image
	const imageStyle = [
		BR.Utils.createStyleSheet({ width: 100 }),
		BR.Utils.createStyleSheet(props.imageStyle || {}),
	];

	let ImageC = ImageComponent || null;
	console.log(ImageC,imageSource);

	if (!ImageC && imageSource) {
		console.log('aaakakakakakakkkkakakakakakkkakakak');
		ImageC = () => (<Image style={imageStyle} source={imageSource} />);
	}

	// Button
	const buttonStyle = [
		BR.Utils.createStyleSheet({ paddingTop: 16 }),
		BR.Utils.createStyleSheet(props.buttonStyle || {}),
	];

	let ButtonC = ButtonComponent || null;
	if (!ButtonC && buttonTitle) {
		ButtonC = () => (<Button style={buttonStyle} onPress={buttonOnPress}><Text>{buttonTitle}</Text></Button>);
	}

	return (
  <View style={stylesheet}>
    {(ImageC) ? <ImageC /> : null}
    {(title) ? <Text style={titleStyle} >{title}</Text> : null}
    {(description) ? <Text style={descriptionStyle} >{description}</Text> : null}
    {(ButtonC) ? <ButtonC /> : null}
  </View>
	);
};

export default withBlueRain(ComponentState);
