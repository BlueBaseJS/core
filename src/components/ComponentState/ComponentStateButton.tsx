import { BlueRainType,withBlueRain } from '../../index';
import { ImageStyle, TextStyle,ViewProperties,ViewStyle } from '@blueeast/bluerain-ui-interfaces';
import React from 'react';

export interface ComponentStateButtonProps extends ViewProperties {
	/**
	 * Button Component, if provided, other button props will be ignored
	 */
	button?: React.ComponentType;

	/**
	 * Button title
	 */
	buttonTitle?: string;

	/**
	 * Button styles
	 */
	buttonStyle?: ViewStyle;

	/**
	 * Button onPress handler
	 */
	buttonOnPress?: Function;

	bluerain: BlueRainType;
}

const ComponentStateButton = (props: ComponentStateButtonProps) => {

	const {
		button: ButtonComponent,
		buttonTitle,
		buttonOnPress,
		buttonStyle,
		bluerain: BR
	} = props;

	if (ButtonComponent) {

		return (<ButtonComponent />);

	} else if (buttonTitle) {
     	console.log(BR);
		const Button = BR.Components.get('Button');
		const Text = BR.Components.get('Text');

		const stylesheet = {
			marginTop: 10,
			...buttonStyle
		};

		return (
			<Button onPress={buttonOnPress} style={BR.Utils.createStyleSheet(stylesheet)}>
				<Text>{buttonTitle}</Text>
			</Button>
		);

	} else {
		return null;
	}
};

export default withBlueRain(ComponentStateButton);
