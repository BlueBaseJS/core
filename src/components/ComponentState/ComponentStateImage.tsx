import { BlueRainType ,withBlueRain } from '../../index';
import { ImageStyle,TextStyle,ViewProperties , ViewStyle } from '@blueeast/bluerain-ui-interfaces';
import React from 'react';

export interface ComponentStateImageProps extends ViewProperties {
	/**
	 * Image Component, if provided, imageSource will be ignored
	 */
	image?: React.ComponentType;

	/**
	 * Image styles
	 */
	imageStyle?: ImageStyle;

	/**
	 * Image source
	 */
	imageSource?: string;

	style?: ViewStyle;

	bluerain: BlueRainType;
}

const ComponentStateImage = (props: ComponentStateImageProps) => {
	const {
		image: ImageComponent,
		imageSource,
		imageStyle,
		bluerain: BR
	} = props;

	const View = BR.Components.get('View');
	const Image = BR.Components.get('Image');

	// Image
	let ImageC;

	if (ImageComponent || imageSource) {
		ImageC = ImageComponent || Image;

	} else {
		return null;
	}

	// Styles
	const stylesheet = { marginBottom: 10, maxWidth: 200 };

	return (
  <View style={BR.Utils.createStyleSheet(stylesheet)}>
		<ImageC
			style={imageStyle ? BR.Utils.createStyleSheet(imageStyle) : {}}
			source={imageSource}
		/>
  </View>
	);
};

export default withBlueRain(ComponentStateImage);
