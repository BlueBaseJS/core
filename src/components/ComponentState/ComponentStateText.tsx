import { BlueRain } from '../../index';
import { TextStyle } from '@blueeast/bluerain-ui-interfaces';
import React from 'react';

export interface ComponentStateTextProps {
	/**
	 * Text
	 */
	text: string;

	/**
	 * Style Object
	 */
	style?: TextStyle;

	bluerain: BlueRain;
}

const ComponentStateText = (props: ComponentStateTextProps) => {

	const {
		text,
		style,
		bluerain: BR
	} = props;

	if (text) {

		const Text = BR.Components.get('Text');

		return (
			<Text style={style ? BR.Utils.createStyleSheet(style, 'Text') : {}} >
				{text}
			</Text>
		);

	} else {
		return null;
	}
};

export default ComponentStateText;
