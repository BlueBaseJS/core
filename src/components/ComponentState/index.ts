import ComponentState from './ComponentState';
export * from './ComponentState';
export default ComponentState;
