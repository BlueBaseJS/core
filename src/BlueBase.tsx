import { Analytics, Logger } from './api';
import {
	ComponentCollection,
	ComponentRegistry,
	ConfigCollection,
	ConfigRegistry,
	HookNestedCollection,
	HookRegistry,
	PluginCollection,
	PluginRegistry,
	ThemeCollection,
	ThemeRegistry,
} from './registries';
import { BlueBaseProvider } from './Context';
import React from 'react';
import { ThemeProvider } from './themes';
import { getComponent } from './getComponent';
import systemHooks from './hooks';

export interface BootOptions {
	components: ComponentCollection,
	configs: ConfigCollection,
	hooks: HookNestedCollection,
	plugins: PluginCollection,
	themes: ThemeCollection,
	// routes: Plugin[]
}

export class BlueBase {

	// APIs
	public Analytics = new Analytics(this);
	public Logger = new Logger(this);

	// Registries
	public Components = new ComponentRegistry(this);
	public Configs = new ConfigRegistry(this);
	public Hooks = new HookRegistry(this);
	public Plugins = new PluginRegistry(this);
	public Themes = new ThemeRegistry(this);

	// Flags
	public booted = false;

	private bootOptions: BootOptions = {
		components: {},
		configs: {},
		hooks: {},
		plugins: [],
		themes: [],
	};

	public async boot(options?: Partial<BootOptions> & { children?: React.ReactNode }) {

		this.bootOptions = { ...this.bootOptions, ...options };

		await this.Hooks.registerNestedCollection(systemHooks);
		await this.Hooks.registerNestedCollection(this.bootOptions.hooks);

		await this.Hooks.run('bluebase.boot', this.bootOptions);

		const SystemApp = getComponent('SystemApp');

		// Set View
		// const SystemApp = this.Components.resolve('SystemApp');
		// SystemApp = await this.Hooks.run('bluebase.system.app', SystemApp);

		const BlueBaseRoot = () => (
			<BlueBaseProvider value={this}>
				<ThemeProvider>
					<SystemApp>
						{options && options.children}
					</SystemApp>
				</ThemeProvider>
			</BlueBaseProvider>
		);

		this.booted = true;

		return BlueBaseRoot;
	}
}