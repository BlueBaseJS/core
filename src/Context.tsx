import { BlueBase } from './BlueBase';
import { createContext } from 'react';

export const BlueBaseContext: React.Context<BlueBase> = createContext(undefined as any);
export const BlueBaseProvider = BlueBaseContext.Provider;
export const BlueBaseConsumer = BlueBaseContext.Consumer;
