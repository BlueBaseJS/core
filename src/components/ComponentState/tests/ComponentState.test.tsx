import * as Native from '../../../native';
import { BlueBase } from '../../../BlueBase';
import { BlueBaseProvider } from '../../../Context';
import { ComponentState } from '../ComponentState';
import React from 'react';
import TestRenderer from 'react-test-renderer';

beforeEach(() => {
	jest.resetModules();
});

const BB = new BlueBase();
BB.Components.register('View', Native.View);
BB.Components.register('Text', Native.Text);
BB.Components.register('Image', Native.Image);



describe('ComponentState', () => {
	const ComponentStateWithProvider = (props: any) => (
		<BlueBaseProvider value={BB}>
			<ComponentState {...props}/>
		</BlueBaseProvider>
	);

	test(`Snapshot ComponentState component`, () => {
		const component = TestRenderer.create(
			<ComponentStateWithProvider />
		);
		const tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});

	test(`Snapshot ComponentState component with description`, () => {
		const component = TestRenderer.create(
			<ComponentStateWithProvider description={'This is just for test'} />
		);
		const tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});

	test(`Snapshot ComponentState component with description and image`, () => {
		const component = TestRenderer.create(
			<ComponentStateWithProvider description={'This is just for test'}
				image={<Native.Image
					style={{ width: 50, height: 50 }}
					source={{ uri: 'https://facebook.github.io/react-native/docs/assets/favicon.png' }}
				/>} />
		);
		const tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});

	test(`Snapshot ComponentState component with description and imageSource`, () => {
		const component = TestRenderer.create(
			<ComponentStateWithProvider description={'This is just for test'}
				imageSource={'hello'} />
		);
		const tree: any = component.toJSON();
		expect(tree.children[0].props.source).toEqual('hello');
	});

});
