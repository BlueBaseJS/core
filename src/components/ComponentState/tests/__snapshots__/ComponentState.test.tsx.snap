// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComponentState Snapshot ComponentState component 1`] = `<View />`;

exports[`ComponentState Snapshot ComponentState component with description 1`] = `
<View>
  <Text>
    This is just for test
  </Text>
</View>
`;

exports[`ComponentState Snapshot ComponentState component with description and image 1`] = `
<View>
  <Image
    source={
      Object {
        "uri": "https://facebook.github.io/react-native/docs/assets/favicon.png",
      }
    }
    style={
      Object {
        "height": 50,
        "width": 50,
      }
    }
  />
  <Text>
    This is just for test
  </Text>
</View>
`;
